### ISSUE TITLE
Solver occasionally does not converge(?) so that the workflow does not start

### ISSUE THREAD

**Issue by fischuu** [snakemake/snakemake#871]
I have a rather complex workflow with 750 samples and roughly 18.000 jobs, at first snakemake runs just fine but then after around 4.000 jobs it suddenly freezes and upon restart it hangs with "Select jobs to execute..." for 24h, after that I terminated it. The initial DAG building takes roughly 2-3 minutes, though.

When I run snakemake (v5.32.0 and v5.32.1) with the `--verbose` option, I get tons of lines similar to this one:

`Cbc0010I After 600 nodes, 304 on tree, -52534.791 best solution, best possible -52538.194 (7.08 seconds`

I tried to delete the `.snakemake` folder in the hope that something went riot there, but that wasn't the case, unfortunately. To me it seems that the CBC MILP Solver somehow does not converge, and it keeps going and going to bring the *best* and the *best possible* solution closer together!? 

After terminating a (shorter) run, I get this verbose output
```
Result - User ctrl-cuser ctrl-c

Objective value:                52534.79114334
Upper bound:                    52538.202
Gap:                            -0.00
Enumerated nodes:               186926
Total iterations:               1807277
Time (CPU seconds):             1181.97
Time (Wallclock seconds):       1188.11
```

I believe it is a bug in the current version, I downgraded Snakemake back to version 5.24, it created then the DAG within 10 minutes and started to run the pipeline. So, apparently there is some bug with the latest version?! 

Unfortunately I cannot create a minimal example to reproduce the problem, as it just happens with this dataset on which I run the pipeline (with a smaller dataset it runs just fine)

**Comment by nick-youngblut** [comment]
I'm getting the same thing with snakemake 5.31.1 running on Ubuntu 18.04.5. The full set of jobs is ~10k. After restart, there are still ~1k jobs left to run. The DAG takes <1 min to construct, but snakemake then seems to hang on `Select jobs to execute..` (hanging for at least 24 hours). 

Verbose output shows at lot of the following:

```
Cbc0010I After 53000 nodes, 14632 on tree, -15341.274 best solution, best possible -15358.966 (151.98 seconds)
Cbc0010I After 54000 nodes, 14743 on tree, -15341.274 best solution, best possible -15358.966 (153.73 seconds)
Cbc0010I After 55000 nodes, 15233 on tree, -15341.274 best solution, best possible -15358.966 (155.08 seconds)
Cbc0010I After 56000 nodes, 15727 on tree, -15341.274 best solution, best possible -15358.966 (156.49 seconds)
Cbc0010I After 57000 nodes, 15752 on tree, -15341.274 best solution, best possible -15358.966 (160.41 seconds)
```

...but never actually starts running any jobs. It appears that the number of nodes, as stated in the verbose output, is infinite.

**Comment by nick-youngblut** [comment]
Using `--scheduler-ilp-solver GLPK_CMD` doesn't seem to help. I instead get a lot of the following when using `--verbose`:

```
+ 12867: >>>>>   1.534127403e+04 <=   1.535908521e+04   0.1% (312; 456)
+ 17804: mip =   1.534127403e+04 <=   1.535907704e+04   0.1% (620; 469)
+ 22704: mip =   1.534127403e+04 <=   1.535907178e+04   0.1% (1005; 479)
+ 27444: mip =   1.534127403e+04 <=   1.535906649e+04   0.1% (1758; 489)
+ 32500: mip =   1.534127403e+04 <=   1.535906480e+04   0.1% (2239; 500)
+ 37372: mip =   1.534127403e+04 <=   1.535906180e+04   0.1% (2999; 511)
+ 42202: mip =   1.534127403e+04 <=   1.535905805e+04   0.1% (3685; 521)
+ 46942: mip =   1.534127403e+04 <=   1.535905588e+04   0.1% (4467; 531)
+ 51874: mip =   1.534127403e+04 <=   1.535905294e+04   0.1% (5246; 542)
+ 56855: mip =   1.534127403e+04 <=   1.535905113e+04   0.1% (5996; 552)
+ 61871: mip =   1.534127403e+04 <=   1.535904903e+04   0.1% (6689; 563)
Time used: 60.0 secs.  Memory used: 8.9 Mb.
+ 66792: mip =   1.534127403e+04 <=   1.535904748e+04   0.1% (7498; 573)
+ 71881: mip =   1.534127403e+04 <=   1.535904493e+04   0.1% (8105; 584)
+ 76715: mip =   1.534127403e+04 <=   1.535904351e+04   0.1% (9138; 595)
```

**Comment by nick-youngblut** [comment]
This hanging issue does NOT happen when I simply use `snakemake -j 12 --use-conda --configfile ./path/to/config.yaml`. The issue only occurs when using my standard bash script for running snakemake pipelines:

```
#!/bin/bash

# user input
if [ "$#" -lt 2 ]; then
    echo "snakemake_sge.sh config.yaml jobs ..."
    echo " config.yaml : snakemake config"
    echo " jobs : number of parallel qsub jobs"
    echo " ... : additional arguments passed to snakemake"
    exit
fi

# check for snakemake
command -v snakemake >/dev/null 2>&1 || { echo "snakemake is not in your PATH"; exit 1; }

# set args
CONFIG=$1
JOBS=$2

# snakemake call
WORKDIR=`pwd`
snakemake \
    --conda-frontend mamba \
    --profile bin/ll_pipeline_utils/profiles/sge/ \
    --use-conda \
    --configfile $CONFIG \
    --jobs $JOBS \
    --local-cores $JOBS \
    --printshellcmds \
    --resources temp=$JOBS \
    --directory $WORKDIR \
    "${@:3}"
```

**Comment by nick-youngblut** [comment]
Even when removing `--conda-frontend`, `--local-cores`, `--printshellcmds`, `--resources` and `--directory`, calling snakemake via the bash script still results in snakemake hanging at `Select jobs to execute...`. 

It might be the SGE profile. That can be found at: https://github.com/leylabmpi/ll_pipeline_utils/tree/master/profiles/sge

**Comment by nick-youngblut** [comment]
At least for me, my pipeline has multiple independent sections, so I was able to just run one portion of the pipeline at a time, and in this case snakemake did not hang, even when calling snakemake via my bash script.

**Comment by geparada** [comment]
Hello there,
I encountered a similar issue running a big workflow of ~100K steps. I my case snakemake stop submitting jobs after ~1K jobs and then get stuck displaying `Select jobs to execute..` 

Then I found @fischuu question on stackoverflow:

https://stackoverflow.com/questions/66122961/select-jobs-to-execute-runs-literally-forever

And one of the recent answers helped me to get over this. Basically just using `--scheduler greedy` did the job for me, but this makes me wonder why the IPL solver get stuck on this. Could @johanneskoester give us any hint of why the IPL solver get stuck some times? what are the advantages of one solver over the other? Is it always the case that the greedy solver is better for large workflows?? 

Cheers!

**Comment by jeskowagner** [comment]
Just wanted to chime in that in my use-case, running 150k jobs, the same issue occurs and is also fixed by `--schedule greedy`. Without it, similar to what @geparada describes, many jobs are submitted and run fine, but eventually I get thousands of lines of `Select jobs to execute..`, which does not resolve overnight.

**Comment by johanneskoester** [comment]
Fixed with PR #2938
